lib/at_cmd_parser/at_cmd_parser.c

```diff
@@ -367,6 +367,91 @@ static int at_parse_param(const char **at_params_str,
 	return 0;
 }
 
+/*
+ * Internal function.
+ * Check AT command grammar based on below.
+ *  AT<terminator>
+ *  AT<separator><body><terminator>
+ *  AT<separator><body>=<terminator>
+ *  AT<separator><body>?<terminator>
+ *  AT<separator><body>=?<terminator>
+ *  AT<separator><body>=<parameters><terminator>
+ * In which
+ * <terminator>: CR, LF, CRLF, NULL
+ * <separator>: +, %, #
+ * <body>: alphabetic char only, size > 0
+ * <parameters>: arbitrary, size > 0
+ */
+int command_grammar_check(const char *at_params_str)
+{
+	const char *cmd = at_params_str;
+	const char *body = NULL;
+
+	/* check AT (if not, no check) */
+	if (strlen(cmd) < 2 || toupper((int)cmd[0]) != 'A' || toupper((int)cmd[1]) != 'T') {
+		return 0;
+	}
```

Comment:
Would be good to add an empty line after the closing braces, currently it's a bit cramped.